### PR TITLE
[structured config] Base structured config implementation

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Introduces the capability to specify config types for an Op or Asset by applying a type annotation to the new `config` keyword, e.g.

```python
@op
def greet_user(config: str):
    print(f"Hello, {config}")
```

For more complex config schema, users can build a Pydantic type ("structured config") extending the new `Config` base class. This provides type hinting when using config in an asset or op body:

```python
class GreetingConfig(Config):
    name: str
    should_print: bool

@op 
def greet_user(config: GreetingConfig):
    if config.should_print:
        print(f"Hello, {config.name}")
    else:
        ...
```

## Limitations

This PR only handles the case of structured config which has primitive types. More complex config (defaults, descriptions, Permissive, Selector/Unions, etc) will be handled in stacked diffs.

## Test Plan

Introduced a set of unit tests to cover basic config classes using primitives + nested structured config.

### CODE REVIEW

**Review by schrockn — DISMISSED** [review]
See inline comments

File: python_modules/dagster/dagster/_config/structured_config.py
Comment (schrockn) [inline python_modules/dagster/dagster/_config/structured_config.py:48]: This is pretty odd.

1) If the type annotations are to be trusted, issubclass should always succeed.
2) Even though some say it is "Pythonic" I think this is a dangerous coding pattern, as it is difficult to know the universe of reasons why something will throw an error.

I suggest instead checking `isinstance(cls, Type)` and returning False in that case.

File: python_modules/dagster/dagster/_config/structured_config.py
Comment (schrockn) [inline python_modules/dagster/dagster/_config/structured_config.py:55]: you could also do an `inst` check here

File: python_modules/dagster/tox.ini
Comment (schrockn) [inline python_modules/dagster/tox.ini:45]: can you comment as to why this is necessary

**Review by schrockn — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/_config/structured_config.py
Comment (schrockn) [reply python_modules/dagster/dagster/_config/structured_config.py:55]: `check.inst_param(cls, Type)` to be precise

File: python_modules/dagster/dagster/_config/structured_config.py
Comment (benpankow) [reply python_modules/dagster/dagster/_config/structured_config.py:48]: Will gate with an `isinstance(cls, Type)` check, as suggested - "issubclass should always succeed" is true if `cls` is a class, but in this case we are occasionally dealing with non-class objects (e.g. `typing` annotations, `issubclass(List[str], Config)` raises a `TypeError`).

File: python_modules/dagster/tox.ini
Comment (benpankow) [reply python_modules/dagster/tox.ini:45]: This is because of a pylint issue, this flag [allows pylint to recognize various top-level imports](https://github.com/pydantic/pydantic/issues/1961#issuecomment-729288794) from the `pylint` module. Added a comment in the `.ini` file as well.

**Review by schrockn — APPROVED** [review]
Awesome

**Comment by benpankow** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11268?utm_source=stack-comment).


Job Reference: TXP57Jx30eGCn07R3Kfh

**Pull request merged**
